File: fast_sync_dotnet_service/fast-sync-core/abstraction/models/SyncOperationMetadata.cs
using fast_sync_core.implementation.metadata;

namespace fast_sync_core.abstraction.data
{
    public class SyncOperationMetadata
    {
        public Dictionary<string, SyncMetadata> Metadata { get; private set; }

        public SyncOperationMetadata()
        {
            Metadata = new Dictionary<string, SyncMetadata>();
        }

        public static SyncOperationMetadata Create(SyncOperationMetadata operationMetadata)
        {
            var syncOperationMetadata = new SyncOperationMetadata();
            foreach (var keyValuePair in operationMetadata.Metadata)
            {
                var value = keyValuePair.Value;

                SyncMetadata syncMetadata = new SyncMetadata();
                syncMetadata.Id = value.Id;
                syncMetadata.SyncZone = value.SyncZone;
                syncMetadata.Version = value.Version;
                syncMetadata.Type = value.Type;
                syncMetadata.SyncOperation = value.SyncOperation;
                syncMetadata.Timestamp = value.Timestamp;                
                syncOperationMetadata.SetMetadata(keyValuePair.Key, syncMetadata);
            }

            return syncOperationMetadata;
        }

        public void SetMetadata(string type, SyncMetadata metadata)
        {
            Metadata[type] = metadata;
        }

        public SyncMetadata GetTypeMetadata(string type)
        {
            if (Metadata.ContainsKey(type) && Metadata[type] != null)
            {
                return Metadata[type];
            }
            throw new InvalidOperationException("Metadata of each synced type should be specified, please check how you build SyncOperationMetadata");
        }

        public List<string> GetSyncedTypes()
        {
            return Metadata.Keys.ToList();
        }
    }
}

File: fast-sync-typescript-client-sdk/packages/fast-sync-client/src/abstraction/models/sync_operation_metadata.ts
import { SyncMetadata } from "../../implementation/metadata/syncable_metadata";
import { isNullOrUndefined } from "../../implementation/utils/utils";
import { ISyncMetadata } from "../metadata/ISync_metadata";

export class SyncOperationMetadata {
  metadata: Record<string, ISyncMetadata> = {};
  constructor() {}

  static create(operationMetadata: SyncOperationMetadata) {
    let syncOperationMetadata = new SyncOperationMetadata();
    const keys = Object.keys(operationMetadata.metadata);
    for (const key of keys) {
      const value = operationMetadata.metadata[key];
      let syncMetadata = new SyncMetadata(
        value.type,
        value.version,
        value.syncZone
      );
      syncMetadata.id = value.id;
      syncMetadata.syncOperation = value.syncOperation;
      syncMetadata.timestamp = value.timestamp;
      syncOperationMetadata.setMetadata(key, syncMetadata);
    }

    return syncOperationMetadata;
  }

  public setMetadata(type: string, metadata: ISyncMetadata): void {
    this.metadata[type] = metadata;
  }

  public getTypeMetadata(type: string): ISyncMetadata {
    if (
      this.metadata.hasOwnProperty(type) &&
      !isNullOrUndefined(this.metadata[type])
    ) {
      return this.metadata[type];
    }
    throw new Error(
      "metadata of each syncked type should specified, please check how you build SyncOperationMetadata"
    );
  }

  public getSynckedTypes(): string[] {
    return Object.keys(this.metadata);
  }
}
